### PR TITLE
Search coordinates

### PR DESCRIPTION
**Pull request by ger-benjamin**
Ref to: https://github.com/camptocamp/ngeo/issues/696 and https://github.com/camptocamp/c2cgeoportal/issues/2044
Specs:  https://github.com/camptocamp/c2cgeoportal/wiki/Spec-%23696-Search-for-coordinates-with-marker

Example: https://ger-benjamin.github.io/ngeo/search_coordinates/examples/contribs/gmf/apps/mobile

### CODE REVIEW

**Comment by fredj** [comment]
Instead of having `exports/services/autoprojection.js` add a `@export` to the methods you want to export

File: src/services/autoprojection.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+ * TODO
+ * @param {ol.Coordinate} coordinates TODO
+ * @param {ol.proj.Projection} projection TODO
+ * @param {Array.<ol.proj.Projection>=} opt_projections TODO
+ * @return {?ol.Coordinate} TODO
+ * @export
+ */
+ngeo.AutoProjection.prototype.tryProjections = function(coordinates, projection,
+    opt_projections) {
+  var extent = projection.getExtent();
```
Comment (sbrunner) [inline src/services/autoprojection.js]: I think that we should try to get the extent from the view ...

File: src/services/autoprojection.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+ngeo.AutoProjection = function() {};
+
+
+/**
+ * TODO
+ * @param {string} str TODO
+ * @return {?ol.Coordinate} TODO
+ * @export
+ */
+ngeo.AutoProjection.prototype.stringToCoordinates = function(str) {
```
Comment (sbrunner) [inline src/services/autoprojection.js]: If str is `' '` I can't predict that the code will do...
I think that if one of the `parseFloat` return `NaN`, this function should return `null`.

Comment (tsauerwein) [inline src/services/autoprojection.js]: It would be good to have unit-tests for that.

**Comment by ger-benjamin** [comment]
Normally, the style used for the marker must be customizable but I think that this point must be in another PR / issue. Because I think that if we be able to customize the "coordinates" marker, we can be able to customize the "normal search" style too.

**Comment by sbrunner** [comment]
We should use the extent set on the view if it's define, if not we will use the projection extent :-)
We don't want to recenter on a point that not possible to use as a center ...
Thanks for the rename :-)

**Comment by ger-benjamin** [comment]
Done. I've added a "extent" parameter. More parameters but this is more logical at the end I think.

File: src/services/autoprojection.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+ngeo.AutoProjection.prototype.tryProjections = function(coordinates,
+    extent, viewProjection, opt_projections) {
+  var position;
+  if (opt_projections === undefined) {
+    opt_projections = [viewProjection];
+  }
+  opt_projections.some(function(projection) {
+    position = ol.proj.transform(coordinates, projection, viewProjection);
+    if (ol.extent.containsCoordinate(extent, position)) {
+      return true;
```
Comment (sbrunner) [inline src/services/autoprojection.js]: not `return projection`?

File: src/services/autoprojection.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    }
+    position = null;
+  });
+  return position;
```
Comment (sbrunner) [inline src/services/autoprojection.js]: can just be `return null`...

File: contribs/gmf/apps/desktop/index.html [hunk 5]
```diff
@@ -87,6 +87,7 @@
           <gmf-search gmf-search-map="mainCtrl.map"
             gmf-search-datasources="mainCtrl.searchDatasources"
             gmf-search-currenttheme="mainCtrl.theme"
+            gmf-search-coordinatesprojections="'EPSG:21781,EPSG:2056,EPSG:4326'"
```
Comment (sbrunner) [inline contribs/gmf/apps/desktop/index.html]: Shouldn't be in the abstractController?

File: src/services/autoprojection.js (hunk 3 above)
Comment (ger-benjamin) [inline src/services/autoprojection.js]: No, that's in a "some" function. I use `return true` as a `break` in my loop.

File: src/services/autoprojection.js (hunk 4 above)
Comment (ger-benjamin) [inline src/services/autoprojection.js]: No, that's in a "some" function. I use `return true` as a `break` in my loop.

File: contribs/gmf/src/directives/search.js [hunk 6]
```diff
@@ -3,17 +3,22 @@ goog.provide('gmf.searchDirective');
...
 goog.require('ngeo.FeatureOverlayMgr');
+goog.require('ol.Feature');
+goog.require('ol.geom.Point');
 /**
  * This goog.require is needed because it provides 'ngeo-search' used in
  * the template.
  * @suppress {extraRequire}
  */
 goog.require('ngeo.searchDirective');
+//goog.require('ol.Feature');
```
Comment (sbrunner) [inline contribs/gmf/src/directives/search.js]: should be removed

File: contribs/gmf/src/directives/search.js [hunk 7]
```diff
@@ -3,17 +3,22 @@ goog.provide('gmf.searchDirective');
...
 goog.require('ol.Map');
 goog.require('ol.proj');
+//goog.require('ol.geom.Point');
```
Comment (sbrunner) [inline contribs/gmf/src/directives/search.js]: also

**Comment by sbrunner** [comment]
Two comments to remove also that's looks good for me :-)

**Comment by tsauerwein** [comment]
Thanks for adding the tests!

**Comment by sbrunner** [comment]
@tsauerwein +1

**Comment by sbrunner** [comment]
Thanks, merge :-)

**Pull request merged**